### ISSUE TITLE
Spec local cache

### ISSUE THREAD

**Issue by michaelsena** [uport-project/3box-js#171]

**Comment by zachferland** [comment]
First task is creating a simple setup for testing, need to try some things out. Not too much to add yet. 

As oed mentioned elsewhere https://github.com/uport-project/3box/issues/219#issuecomment-428454491 this is the basic idea and would need to see if and what parts of the orbited api could we wrap, likely some RPC wrapper would make sense. 

Alternatively I want to see if we could just run orbit db in both window and iframe and connect them as peers through iframe message passing. Pros: same libs, more stable than other wrappers, has a good chance of being less effort/code. Cons: loading ipfs libs

**Comment by oed** [comment]
Oh, I've thought about that too. Would that require one ipfs node in the iframe and one in the lib? Do you think it's easy to make the two nodes connect?
Loading the ipfs lib twice seems like a lot though :/

**Comment by zachferland** [comment]
yeah one in iframe and one in libs. Not sure how hard or easy right now, thats one of the immediate questions. Don't think super trivial, likely more than just some configs. Actually not really worried about loading libs twice if ipfs pulled out of lib, then they would likely just be fetching the same ipfs js file from some CDN and would be cached in browser anyways.

**Comment by oed** [comment]
ah, ok. Makes sense. Hopefully we can somehow affect the cache of that file also :)

**Comment by zachferland** [comment]
https://github.com/ipfs/js-ipfs#optionslibp2p
https://github.com/libp2p/js-libp2p
https://github.com/libp2p/interface-transport
https://github.com/ipfs/js-ipfs/tree/master/examples/custom-libp2p

**Comment by zachferland** [comment]
Creating a proxy lib wrapper is not too trivial, there is considerable overhead (for example size of this codebase for ipfs one, then api changes, issues, etc). The overhead is high at least for libs with high api surface area (i.e. orbitdb, and ones with hard to serialize args/returns). Related there is the option to wrap our entire lib and move most all functionality to iframe, but that comes with considerable cons. If we just wrap top level api, then developer loses access to all nested interfaces (like stuff from orbitdb), and can only use surface level api (limit experimentation, and likely limited use cases). Also rather restrictive in the future. Basically the conclusion is we need to find a lower level api/layer with minimal surface area to pass message through. 

So two paths remain: 

1) Find a lower level, minimal api, in orbit db to insert an message passing interface between window and iframe. Ideally one that is easy to serialize as well. 
2) reframe problem and instead create a libp2p module for transport over post message, and simply connect both instances as peers (orbitdb and ipfs). 

(cc @oed )

**Comment by zachferland** [comment]
Ref 1: wrapping just the orbit-db-cache lib https://github.com/orbitdb/orbit-db-cache/blob/master/Cache.js which implements the local cache might the right solution

**Comment by oed** [comment]
Right, yeah that seems like a lot of overhead. Path 1 seems interesting. I'm wondering though, does orbitdb store all of the events in this cache or only stuff like most recent head etc? I.e. is the ipfs storage actually used? Feel free to join the orbitdb gitter and ask their dev team about this though. We can also raise the question during next weeks orbitdb community call.

Re path 2: If I understand you correctly we run one ipfs+orbitdb instance in the iframe and one ipfs+orbitdb instance in the library. I think these nodes might be able to find each other without any additional code, but we could probably speed it up by connecting them directly.

**Comment by oed** [comment]
Just asked about path 1 in the orbitdb gitter:


---
Joel Torstensson
@oed
10:20
What data is orbitdb storing in it's cache, i.e. not in ipfs?
We are researching ways to share orbitdb data between multiple domains using an iframe :)
Haad
@haadcode
10:21
@oed the "cache" (as in orbit-db-cache and the local ./orbitdb data directory) is a leveldb that saves (per-db) the latest heads (iirc we save the full entry and not just the hash of the entry, although I believe there's no reason why it couldn't be just the hash too, long time since I touched that part), the latest "remote heads" (latest ones received from the network, whereas "heads" is latest of either from network or locally) and the hash of the db manifest file.
none other than the "heads" is needed for functionality, all others are more or less optimizations or convenience
Joel Torstensson
@oed
10:27
@haadcode ok cool. So when you open a store you basically: get the latest heads from the cache, then get the rest of the data from ipfs?
Just making sure that I understand it correctly!
Haad
@haadcode
10:58
@oed: yeah, pretty much. you get the latest hash from cache so you can open the "persisted state" of the db (and get going), and anything that come from the network can be merged later (as it can be newer/older entries that you receive from peers).

--- 

So if going with path 1 we still need a way to share the ipfs storage. Potentially this could be fetched from the ipfs network, but that will add some latancy.

**Comment by zachferland** [comment]
1) I was thinking of using the ipfs proxy along with a proxy for the orbitdb cache. I imagine it could work, but haven't thought it through yet and/or just tried it. 

2) Yeah I they should be able to find each other, there would be some roundtrip time to find each other, but the communication channel may be a short path after that (especially webrtc on same machine), but have to look closer to see what would be gained. But yeah I think the most gain would be if they can communicate directly over postMessage using a libp2p module as an abstraction and directly find each other without any discovery services.

**Comment by oed** [comment]
Cool!
To me it seems like 1 would be the best option. https://github.com/ipfs-shipyard/ipfs-postmsg-proxy looks like it's actively maintained. Would require a PR to orbitdb to allow it to use a custom cache though. 

You might have more insight into 2 though. You think that might be better in some other way?
